File: sdks/aperture-js/example/concurrency_scheduler.ts
import {ApertureClient} from "@fluxninja/aperture-js";

import inquirer from 'inquirer';

async function initializeApertureClient() {
    const answers = await inquirer.prompt([
        {
            type: 'input',
            name: 'address',
            message: 'Enter your organization\'s address:',
        },
        {
            type: 'input',
            name: 'apiKey',
            message: 'Enter the API key:',
        },
    ]);

    const apertureClient = new ApertureClient({
        address: answers.address,
        apiKey: answers.apiKey,
    });

    return apertureClient;
}

// Define user tiers and associated priorities
const userTiers = {
  platinum: 8,
  gold: 4,
  silver: 2,
  free: 1,
};

async function sendRequestForTier(
  apertureClient: ApertureClient,
  tier: string,
  priority: number,
) {
  console.log(`[${tier} Tier] Sending request with priority ${priority}...`);
  // START: CSStartFlow
  // START: UICS
  const flow = await apertureClient.startFlow(
    "concurrency-scheduling-feature",
    {
      labels: {
        user_id: "some_user_id",
        priority: priority.toString(),
        workload: tier,
      },
      grpcCallOptions: {
        deadline: Date.now() + 120000, // ms
      },
    },
  );
  // END: UICS

  if (flow.shouldRun()) {
    console.log(`[${tier} Tier] Request accepted with priority ${priority}.`);
    // sleep for 5 seconds to simulate a long-running request
    await new Promise((resolve) => setTimeout(resolve, 5000));
  } else {
    console.log(`[${tier} Tier] Request rejected. Priority was ${priority}.`);
  }

  flow.end();
  // END: CSStartFlow
}

// Launch each batch in parallel
async function scheduleRequests(apertureClient: ApertureClient) {
  const requestsPerBatch = 10;
  const batchInterval = 1000; // ms

  while (true) {
    console.log("Sending new batch of requests...");
    // Send requests for each tier
    const promises = Object.entries(userTiers).flatMap(([tier, priority]) => {
      return Array(requestsPerBatch)
        .fill(null)
        .map(() => sendRequestForTier(apertureClient, tier, priority));
    });

    await Promise.all(promises);
    await new Promise((resolve) => setTimeout(resolve, batchInterval));
  }
}

async function main() {
  const apertureClient = await initializeApertureClient();
  scheduleRequests(apertureClient);
}

main().catch((e) => console.error(e));

File: sdks/aperture-py/example/concurrency_scheduler.py
import asyncio
from datetime import timedelta

from aperture_sdk.client_async import ApertureClientAsync, FlowParams
from aperture_sdk.flow_common import FlowStatus

user_tiers = {
    "free": 1,
    "silver": 2,
    "gold": 4,
    "platinum": 8,
}
requests_per_batch = 50
batch_interval = 1

agent_address = ""
api_key = ""


def initialize_aperture_client():
    aperture_client = ApertureClientAsync.new_client(
        address=agent_address, api_key=api_key
    )
    return aperture_client


async def send_request_for_tier(i, aperture_client, tier, priority):
    print(f"[{tier} Tier] Sending request with priority {priority}...")
    flow_params = FlowParams(
        explicit_labels={
            "priority": str(priority),
            "workload": tier,
        },
        check_timeout=timedelta(seconds=10),
    )

    flow = await aperture_client.start_flow(
        control_point="concurrency-scheduling-feature",
        params=flow_params,
    )

    if flow.should_run():
        await asyncio.sleep(0.1)
        print(f"[{tier} Tier] Request accepted. Priority was {priority}.")
    else:
        flow.set_status(FlowStatus.Error)
        print(f"[{tier} Tier] Request rejected. Priority was {priority}.")

    await flow.end()


async def schedule_requests(aperture_client):
    while True:
        print("Sending new batch of requests...")

        tasks = []
        for tier, priority in user_tiers.items():
            for i in range(requests_per_batch):
                task = asyncio.create_task(
                    send_request_for_tier(i, aperture_client, tier, priority)
                )
                tasks.append(task)

        await asyncio.gather(*tasks)
        await asyncio.sleep(batch_interval)


async def main():
    aperture_client = initialize_aperture_client()
    await schedule_requests(aperture_client)


if __name__ == "__main__":
    asyncio.run(main())
